File: lib/src/replay/replay_fingerprint.dart
import '../determinism/fingerprint.dart';
import '../graph/runtime_graph.dart';
import 'replay_graph.dart';

Map<String, dynamic> validateFingerprintReplayEquivalence(
  Map<String, dynamic> original,
  Map<String, dynamic> replayed, {
  RuntimeGraph? graph,
  Map<String, dynamic>? memory,
}) {
  final g = graph ?? _graphFromEnvelope(original);
  final rg = _graphFromEnvelope(replayed);
  final origFp = computeGlobalRuntimeFingerprint(original, g);
  final replayFp = computeGlobalRuntimeFingerprint(replayed, rg);
  final graphHashMatch = graphReplayHash(g) == graphReplayHash(rg);
  return {
    'equivalent': origFp == replayFp && graphHashMatch,
    'global_fingerprint_match': origFp == replayFp,
    'graph_hash_match': graphHashMatch,
    'bounded': true,
  };
}

RuntimeGraph _graphFromEnvelope(Map<String, dynamic> envelope) {
  final g = envelope['unified_runtime_graph'] ?? envelope['graph'];
  if (g is RuntimeGraph) return g;
  if (g is Map<String, dynamic>) {
    return buildRuntimeGraph({'graph': g});
  }
  return RuntimeGraph(nodes: [], edges: []);
}

File: src/replay/replayFingerprint.ts
import { computeGlobalRuntimeFingerprint } from "../determinism/globalRuntimeFingerprint.js";
import type { ExtractionEnvelope } from "../contracts/runtimeContracts.js";
import type { RuntimeGraph } from "../contracts/graphContracts.js";
import { graphReplayHash } from "./replayGraph.js";

export function computeReplayFingerprint(
  envelope: ExtractionEnvelope,
  graph?: RuntimeGraph,
  memory?: Record<string, unknown>,
): string {
  return computeGlobalRuntimeFingerprint(envelope, graph, memory);
}

export function validateFingerprintReplayEquivalence(
  original: ExtractionEnvelope,
  replayed: ExtractionEnvelope,
  graph?: RuntimeGraph,
  memory?: Record<string, unknown>,
): {
  equivalent: boolean;
  global_fingerprint_match: boolean;
  graph_hash_match: boolean;
  bounded: boolean;
} {
  const g = graph ?? (original.unified_runtime_graph as RuntimeGraph);
  const rg = (replayed.unified_runtime_graph ?? replayed.graph ?? g) as RuntimeGraph;
  const origFp = computeReplayFingerprint(original, g, memory);
  const replayFp = computeReplayFingerprint(replayed, rg, memory);
  const graph_hash_match = graphReplayHash(g) === graphReplayHash(rg);
  return {
    equivalent: origFp === replayFp && graph_hash_match,
    global_fingerprint_match: origFp === replayFp,
    graph_hash_match,
    bounded: true,
  };
}
